universe/vault/credentials.go

```diff
@@ -48,19 +63,70 @@ func (c *Credentials) Login(ctx context.Context, options ...vault.ClientOption)
 		}
 	}
 
-	resp, err := client.Auth.AppRoleLogin(
+	return &ClientHandle{
+		creds:  c,
+		client: client,
+	}, nil
+}
+
+func (h *ClientHandle) Get(ctx context.Context) (*vault.Client, error) {
+	h.m.Lock()
+	defer h.m.Unlock()
+
+	// Re-authenticate if less than six seconds left.
+	if time.Now().Add(ttlBuffer).After(h.expires()) {
+		return h.client, h.renew(ctx)
+	}
+
+	return h.client, nil
+}
+
+func (h *ClientHandle) authenticate(ctx context.Context) error {
+	resp, err := h.client.Auth.AppRoleLogin(
 		ctx,
 		schema.AppRoleLoginRequest{
-			RoleId:   c.RoleId,
-			SecretId: c.SecretId,
+			RoleId:   h.creds.RoleId,
+			SecretId: h.creds.SecretId,
 		},
 	)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	if err := client.SetToken(resp.Auth.ClientToken); err != nil {
-		return nil, err
+	h.auth = resp.Auth
+	h.leased = time.Now()
+	zerolog.Ctx(ctx).Debug().Dur("lease_duration", h.ttl()).Msg("vault: authenticated")
+	return h.client.SetToken(resp.Auth.ClientToken)
+}
+
+func (h *ClientHandle) renew(ctx context.Context) error {
+	if h.auth == nil || !h.auth.Renewable {
+		return h.authenticate(ctx)
 	}
 
-	return client, nil
+	res, err := h.client.Auth.TokenRenewSelf(ctx, schema.TokenRenewSelfRequest{})
+	if err != nil {
+		// The Vault client library already handles retries, so if renewing the
+		// token fails, we assume it can no longer be renewed. This can happen if
+		// the token was revoked, or if it reached its maximum TTL.
+		zerolog.Ctx(ctx).Warn().Msg("vault: token renewal failed, forcing re-auth")
+		return h.authenticate(ctx)
+	}
+	h.auth = res.Auth
+	zerolog.Ctx(ctx).Debug().Dur("lease_duration", h.ttl()).Msg("vault: token renewed")
+	h.leased = time.Now()
+	return nil
+}
+
+func (h *ClientHandle) ttl() time.Duration {
+	if h.auth == nil {
+		return 0
+	}
+	return time.Duration(h.auth.LeaseDuration) * time.Second
+}
+
+func (h *ClientHandle) expires() time.Time {
```

Comment:
Prefer

```
	if h.auth == nil {
		return time.Time{}
	}

	return h.leased.Add(h.ttl())
```
(new line after block)

Here, and in other places :)
https://www.notion.so/namespacelabs/Style-Guide-331d14027ec64eb0a4bb4a0a3aa4fd07?pvs=4#3355592134f14a9697eedaf1bc35f599

Reply:
Oh we have a style guide :slightly_smiling_face: It seems like a `\n\s*}\n\s*[^}\t]` would match pretty much all bad cases, so maybe we could add a presubmit then?